jmh-generator-asm/src/main/java/org/openjdk/jmh/generators/asm/ASMClassInfo.java

```diff
@@ -79,11 +79,14 @@ public void visit(int version, int access, String name, String signature, String
         this.superName = superName;
         this.idName = name;
         this.access = access;
-        qualifiedName = name.replace("/", ".");
-        packageName = qualifiedName.substring(0, qualifiedName.lastIndexOf("."));
-        origQualifiedName = qualifiedName;
-        qualifiedName = qualifiedName.replace('$', '.');
-        this.name = qualifiedName.substring(qualifiedName.lastIndexOf(".") + 1);
+        this.qualifiedName = name.replace("/", ".");
+
+        int dotIndex = qualifiedName.lastIndexOf(".");
+        if (dotIndex != -1) {
+            packageName = qualifiedName.substring(0, dotIndex);
+        } else {
+            packageName = "";
+        }
```

Comment:
This is not the same as the previous code. Previous code handles "inner" classes with `$` additionally. See my previous comment how this should be resolved. (It is kinda open question if we really need this, but better to keep the semantics the same).

Reply:
D'oh, now I can't even get the copy part of copy/paste right. Selected your entire patch and re-applied this time.